### DISCUSSION TITLE
Gamepad Usage in Android?

### DISCUSSION THREAD

**Discussion by SolarLune** [hajimehoshi/ebiten#1538] [Q&A]
Hello!

I wanted to ask to see how gamepad usage is done in Android? I know the feature [was added awhile back](https://github.com/hajimehoshi/ebiten/issues/1083), so I feel like it should be working, but I can't seem to get it to work. I do have an AndroidManifest.xml file that defines the gamepad as being used, but no ebiten gamepad functions are returning any values. 

I'm using `gomobile build` instead of `ebitenmobile bind` just to get things up and running quickly - could that be the issue?

**Comment by hajimehoshi** [comment]
[go-inovation](https://github.com/hajimehoshi/go-inovation) already worked with gamepads on Android. I don't think I have modified the manifest XML. Have you checked the other application works with gamepads?

**Reply by SolarLune** [reply]
Huh, this is strange. 

I'm testing with two gamepads / setups (one is an 8BitDo SN30 Pro+ paired to my Android phone, a Moto G Power, and the other device is a Retroid Pocket 2, an Android handheld). Both devices recognize when I push inputs on them to navigate through the OS, and games like AM2R, RetroArch running Cave Story, and Sonic Robo Blast 2 detect the gamepad just fine.

However, Inovation and my own test app don't detect any gamepad. I suppose I'll have to take a look to see why the gamepad might not be detected.

As a quick question, does go-inovation work for you with a gamepad on your current device (assuming you currently have an Android phone and a gamepad to test it with)?

EDIT: Ah, and also, I noticed that the Github and docs for gamepad functions mention they always return empty values (for example, 0 or "") for mobile platforms, but I thought that was incorrect and just hasn't been updated, since the missing implementations were added for Android - is this correct?

**Reply by hajimehoshi** [reply]
> As a quick question, does go-inovation work for you with a gamepad on your current device (assuming you currently have an Android phone and a gamepad to test it with)?

Yes. It worked on my Android device with bluetooth gamepads.

> EDIT: Ah, and also, I noticed that the Github and docs for gamepad functions mention they always return empty values (for example, 0 or "") for mobile platforms, but I thought that was incorrect and just hasn't been updated, since the missing implementations were added for Android - is this correct?

Oops, these comments are outdated. I'll update them asap. Thanks!

**Reply by hajimehoshi** [reply]
I tested an Xbox controller. As far as I remember 8bitdo's gamepads require special treatments for Android, but I'm not sure.

**Reply by hajimehoshi** [reply]
Confirmed go-inovation/Pixel4 didn't work with an Xbox controller. Let me investigate more.

**Reply by hajimehoshi** [reply]
Interestingly, sometimes it worked, and sometimes it didn't. ~When it didn't work, the Xbox controller's B button go backs to the top menu from the application go-inovation. That's interesting. I think Ebiten application sometimes fails to get its control.~

**Reply by hajimehoshi** [reply]
Found an issue https://github.com/hajimehoshi/ebiten/issues/1542 and fixed this.

**Reply by hajimehoshi** [reply]
For Retroid Pocket 2, my two cents is that the buttons are detected as keyboard keys rather than gamepad buttons.
